File: 백준/Gold/2294. 동전 2/동전 2.py
import sys
sys.setrecursionlimit(100000)
def input():return sys.stdin.readline().rstrip()
N,M=map(int,input().split())
a=[int(input()) for _ in range(N)]
dp=[-1 for _ in range(M+1)]
def go(pos):
    if pos==M:
        return 0
    if dp[pos]!=-1:
        return dp[pos]
    ret=987654321
    for i in range(N):
        if pos+a[i]<=M:
            ret=min(ret,go(pos+a[i])+1)
    dp[pos]=ret
    return dp[pos]
res=go(0)
if res==987654321:
    res=-1
print(res)

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static int res=0;
    static int N,K;
    static int arr[];
    static int dp[];
    static ArrayList<Integer> v=new ArrayList<>();
    public static int solve(int money){
        if (money==0)
            return 0;
        if (dp[money]!=-1)
            return dp[money];
        int ret=987654321;
        for (int i=0;i<N;i++){
            if (money-v.get(i)>=0){
                ret=Math.min(ret,solve(money-v.get(i))+1);
            }
        }
        dp[money]=ret;
        return dp[money];
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st =new StringTokenizer(br.readLine());
        N=Integer.parseInt(st.nextToken());
        K=Integer.parseInt(st.nextToken());
        dp=new int[K+1];
        Arrays.fill(dp,-1);
        for (int i=0;i<N;i++){
            int num=Integer.parseInt(br.readLine());
            v.add(num);
        }
        res=solve(K);
        if (res==987654321)
            res=-1;
        bw.write(Integer.toString(res));
        bw.close();

    }
}
